File: skills/optimization/reference/src/newton.ts
/**
 * Newton's method for optimization.
 *
 * Uses the Hessian (or its finite-difference approximation) to compute
 * Newton directions d = -H^{-1}*g for quadratic convergence near a minimum.
 * Globalized with line search for robustness far from the solution.
 *
 * @node newton
 * @depends-on vec-ops, result-types, line-search, finite-hessian, finite-diff
 * @contract newton.test.ts
 * @hint modified-newton: When the Hessian is not positive definite, adds
 *       tau*I regularization (Levenberg-Marquardt style) to ensure descent.
 * @hint linear-solve: Uses Cholesky factorization for the linear system.
 *       Falls back to regularization if Cholesky fails.
 * @provenance Nocedal & Wright, Numerical Optimization, Chapter 3 (Newton's method)
 * @provenance Modified Newton: Nocedal & Wright, Section 3.4
 * @provenance Optim.jl Newton() — uses NLSolversBase Hessian infrastructure
 */

import { dot, sub, normInf, addScaled } from "./vec-ops";
import {
  type OptimizeResult,
  type OptimizeOptions,
  defaultOptions,
  checkConvergence,
  isConverged,
  convergenceMessage,
} from "./result-types";
import { wolfeLineSearch } from "./line-search";
import { forwardDiffGradient } from "./finite-diff";
import { finiteDiffHessian } from "./finite-hessian";

/** Options specific to Newton's method. */
export interface NewtonOptions extends Partial<OptimizeOptions> {
  /** Initial regularization for modified Newton. Default: 1e-8 */
  initialTau?: number;
  /** Regularization growth factor. Default: 10 */
  tauFactor?: number;
  /** Maximum regularization attempts. Default: 20 */
  maxRegularize?: number;
}

/**
 * Minimize a function using Newton's method with line search.
 *
 * @param f - Objective function
 * @param x0 - Starting point
 * @param grad - Gradient function (required; if omitted uses finite differences)
 * @param hess - Hessian function (optional; uses finite-diff Hessian if omitted)
 * @param options - Optimization and Newton-specific options
 * @returns OptimizeResult
 *
 * @provenance Nocedal & Wright, Numerical Optimization, Algorithm 3.1 (Newton + line search)
 * @provenance Modified Newton regularization from Section 3.4
 */
export function newton(
  f: (x: number[]) => number,
  x0: number[],
  grad?: (x: number[]) => number[],
  hess?: (x: number[]) => number[][],
  options?: NewtonOptions,
): OptimizeResult {
  const opts = defaultOptions(options);
  const n = x0.length;
  const initialTau = options?.initialTau ?? 1e-8;
  const tauFactor = options?.tauFactor ?? 10;
  const maxRegularize = options?.maxRegularize ?? 20;

  const gradFn = grad ?? ((x: number[]) => forwardDiffGradient(f, x));
  const hessFn = hess ?? ((x: number[]) => finiteDiffHessian(f, x));

  let x = x0.slice();
  let fx = f(x);
  let gx = gradFn(x);
  let functionCalls = 1;
  let gradientCalls = 1;

  // Check if already at minimum
  let gradNorm = normInf(gx);
  const initialCheck = checkConvergence(gradNorm, Infinity, Infinity, 0, opts);
  if (initialCheck && isConverged(initialCheck)) {
    return {
      x, fun: fx, gradient: gx.slice(),
      iterations: 0, functionCalls, gradientCalls,
      converged: true, message: convergenceMessage(initialCheck),
    };
  }

  for (let iteration = 1; iteration <= opts.maxIterations; iteration++) {
    // Compute Hessian
    const H = hessFn(x);

    // Solve H*d = -g using Cholesky, with regularization if needed
    const negG = gx.map(g => -g);
    const d = solveWithRegularization(H, negG, initialTau, tauFactor, maxRegularize);

    if (!d) {
      return {
        x, fun: fx, gradient: gx.slice(),
        iterations: iteration, functionCalls, gradientCalls,
        converged: false,
        message: "Stopped: Hessian regularization failed — cannot find descent direction",
      };
    }

    // Ensure descent direction
    if (dot(d, gx) >= 0) {
      // Fall back to steepest descent
      for (let i = 0; i < n; i++) d[i] = -gx[i];
    }

    // Line search
    const ls = wolfeLineSearch(f, gradFn, x, d, fx, gx);
    functionCalls += ls.functionCalls;
    gradientCalls += ls.gradientCalls;

    if (!ls.success) {
      return {
        x, fun: fx, gradient: gx.slice(),
        iterations: iteration, functionCalls, gradientCalls,
        converged: false,
        message: "Stopped: line search failed to find acceptable step",
      };
    }

    const xNew = addScaled(x, d, ls.alpha);
    const fNew = ls.fNew;
    const gNew = ls.gNew ?? gradFn(xNew);
    if (!ls.gNew) gradientCalls++;

    const sk = sub(xNew, x);
    const stepNorm = normInf(sk);
    const funcChange = Math.abs(fNew - fx);
    gradNorm = normInf(gNew);

    x = xNew;
    fx = fNew;
    gx = gNew;

    const reason = checkConvergence(gradNorm, stepNorm, funcChange, iteration, opts);
    if (reason) {
      return {
        x: x.slice(), fun: fx, gradient: gx.slice(),
        iterations: iteration, functionCalls, gradientCalls,
        converged: isConverged(reason), message: convergenceMessage(reason),
      };
    }
  }

  return {
    x: x.slice(), fun: fx, gradient: gx.slice(),
    iterations: opts.maxIterations, functionCalls, gradientCalls,
    converged: false,
    message: `Stopped: reached maximum iterations (${opts.maxIterations})`,
  };
}

/**
 * Solve H*d = b using Cholesky factorization, with regularization if needed.
 *
 * If Cholesky fails (H not positive definite), adds tau*I and retries.
 * Returns null if regularization is exhausted.
 */
function solveWithRegularization(
  H: number[][],
  b: number[],
  initialTau: number,
  tauFactor: number,
  maxAttempts: number,
): number[] | null {
  const n = b.length;

  // Try Cholesky without regularization first
  const d = choleskySolve(H, b);
  if (d) return d;

  // Add regularization: solve (H + tau*I)*d = b
  let tau = initialTau;
  for (let attempt = 0; attempt < maxAttempts; attempt++) {
    const Hreg: number[][] = Array.from({ length: n }, (_, i) => {
      const row = H[i].slice();
      row[i] += tau;
      return row;
    });

    const dReg = choleskySolve(Hreg, b);
    if (dReg) return dReg;

    tau *= tauFactor;
  }

  return null;
}

/**
 * Solve A*x = b using Cholesky factorization (A = L*L').
 * Returns null if A is not positive definite.
 */
function choleskySolve(A: number[][], b: number[]): number[] | null {
  const n = b.length;

  // Cholesky factorization: A = L * L'
  const L: number[][] = Array.from({ length: n }, () => new Array(n).fill(0));

  for (let i = 0; i < n; i++) {
    for (let j = 0; j <= i; j++) {
      let sum = 0;
      for (let k = 0; k < j; k++) {
        sum += L[i][k] * L[j][k];
      }

      if (i === j) {
        const diag = A[i][i] - sum;
        if (diag <= 0) return null; // Not positive definite
        L[i][j] = Math.sqrt(diag);
      } else {
        L[i][j] = (A[i][j] - sum) / L[j][j];
      }
    }
  }

  // Forward substitution: L*y = b
  const y = new Array(n);
  for (let i = 0; i < n; i++) {
    let sum = 0;
    for (let j = 0; j < i; j++) sum += L[i][j] * y[j];
    y[i] = (b[i] - sum) / L[i][i];
  }

  // Back substitution: L'*x = y
  const x = new Array(n);
  for (let i = n - 1; i >= 0; i--) {
    let sum = 0;
    for (let j = i + 1; j < n; j++) sum += L[j][i] * x[j];
    x[i] = (y[i] - sum) / L[i][i];
  }

  return x;
}

File: research/experiments/sweep/runs/opt-full-python-sonnet45/src/newton.py
"""
Newton's method with Cholesky solve and modified Newton regularization.
"""

import math
from typing import Callable, List, Optional

from .vec_ops import dot, norm_inf, sub, add_scaled, negate
from .result_types import (
    OptimizeResult, default_options, check_convergence,
    is_converged, convergence_message,
)
from .line_search import wolfe_line_search
from .finite_diff import forward_diff_gradient
from .finite_hessian import finite_diff_hessian


def cholesky_solve(A: List[List[float]], b: List[float]) -> Optional[List[float]]:
    """Solve A*x = b via Cholesky. Returns None if not positive definite."""
    n = len(b)
    if n == 0:
        return []
    L = [[0.0] * n for _ in range(n)]

    for i in range(n):
        for j in range(i + 1):
            s = sum(L[i][k] * L[j][k] for k in range(j))
            if i == j:
                diag = A[i][i] - s
                if diag <= 0:
                    return None
                L[i][j] = math.sqrt(diag)
            else:
                L[i][j] = (A[i][j] - s) / L[j][j]

    # Forward substitution: Ly = b
    y = [0.0] * n
    for i in range(n):
        s = sum(L[i][j] * y[j] for j in range(i))
        y[i] = (b[i] - s) / L[i][i]

    # Back substitution: L^T x = y
    x = [0.0] * n
    for i in range(n - 1, -1, -1):
        s = sum(L[j][i] * x[j] for j in range(i + 1, n))
        x[i] = (y[i] - s) / L[i][i]

    return x


def newton(
    f: Callable[[List[float]], float],
    x0: List[float],
    grad: Optional[Callable[[List[float]], List[float]]] = None,
    hess: Optional[Callable[[List[float]], List[List[float]]]] = None,
    grad_tol: float = 1e-8,
    step_tol: float = 1e-8,
    func_tol: float = 1e-12,
    max_iterations: int = 1000,
    initial_tau: float = 1e-8,
    tau_factor: float = 10.0,
    max_regularize: int = 20,
    **kwargs,
) -> OptimizeResult:
    """Minimize using Newton's method with line search."""
    opts = default_options(grad_tol=grad_tol, step_tol=step_tol, func_tol=func_tol,
                           max_iterations=max_iterations)
    grad_fn = grad if grad is not None else (lambda x: forward_diff_gradient(f, x))
    hess_fn = hess if hess is not None else (lambda x: finite_diff_hessian(f, x))

    n = len(x0)
    x = x0[:]
    fx = f(x)
    gx = grad_fn(x)
    function_calls = 1
    gradient_calls = 1

    # Check initial convergence
    grad_norm = norm_inf(gx)
    reason = check_convergence(grad_norm, float('inf'), float('inf'), 0, opts)
    if reason and is_converged(reason):
        return OptimizeResult(
            x=x[:], fun=fx, gradient=gx[:],
            iterations=0, function_calls=function_calls,
            gradient_calls=gradient_calls, converged=True,
            message=convergence_message(reason),
        )

    for iteration in range(1, opts.max_iterations + 1):
        H = hess_fn(x)
        neg_g = [-gi for gi in gx]

        # Try Cholesky with regularization
        d = cholesky_solve(H, neg_g)
        if d is None:
            tau = initial_tau
            for _ in range(max_regularize):
                H_reg = [row[:] for row in H]
                for i in range(n):
                    H_reg[i][i] += tau
                d = cholesky_solve(H_reg, neg_g)
                if d is not None:
                    break
                tau *= tau_factor
            if d is None:
                return OptimizeResult(
                    x=x[:], fun=fx, gradient=gx[:],
                    iterations=iteration, function_calls=function_calls,
                    gradient_calls=gradient_calls, converged=False,
                    message="Stopped: regularization failed",
                )

        # Descent check
        if dot(d, gx) >= 0:
            d = negate(gx)

        ls = wolfe_line_search(f, grad_fn, x, d, fx, gx)
        function_calls += ls.function_calls
        gradient_calls += ls.gradient_calls

        if not ls.success:
            return OptimizeResult(
                x=x[:], fun=fx, gradient=gx[:],
                iterations=iteration, function_calls=function_calls,
                gradient_calls=gradient_calls, converged=False,
                message="Stopped: line search failed",
            )

        x_new = add_scaled(x, d, ls.alpha)
        f_new = ls.f_new
        g_new = ls.g_new if ls.g_new is not None else grad_fn(x_new)
        if ls.g_new is None:
            gradient_calls += 1

        step_norm = norm_inf(sub(x_new, x))
        func_change = abs(fx - f_new)
        grad_norm = norm_inf(g_new)

        x = x_new
        fx = f_new
        gx = g_new

        reason = check_convergence(grad_norm, step_norm, func_change, iteration, opts)
        if reason:
            return OptimizeResult(
                x=x[:], fun=fx, gradient=gx[:],
                iterations=iteration, function_calls=function_calls,
                gradient_calls=gradient_calls,
                converged=is_converged(reason),
                message=convergence_message(reason),
            )

    return OptimizeResult(
        x=x[:], fun=fx, gradient=gx[:],
        iterations=opts.max_iterations, function_calls=function_calls,
        gradient_calls=gradient_calls, converged=False,
        message=f"Stopped: reached maximum iterations ({opts.max_iterations})",
    )
